// @ac4y/ac4y-object/domain/Ac4y.js
var Ac4yAlgebra=require("@ac4y/ac4y-object/algebra/Ac4yAlgebra.js"); 
var Ac4yObjectHandler=require("@ac4y/ac4y-utility/domain/Ac4yObjectHandler.js"); 
var Ac4yJSONHandler=require("@ac4y/ac4y-utility/domain/Ac4yJSONHandler.js");

class Ac4y extends Ac4yAlgebra {

    constructor(object){
        
        super(object);
        if (object!=undefined && object===0) return;
        this.setTemplateName();
    }

    rebuilded(object, target){return super.rebuilded(object, target);}
    
    objectKilledForPersistence(object){

        var object = this.createSelf(object);

        object = new Ac4yObjectHandler().deleteObjectTypeProperties(object);
        
        return object;

    } // objectKilledForPersistence

    killIdentification(){
        delete this.ac4yIdentification;
    }

    onlyGUIDIdentification(object){
        object = object || this;
        delete this.getAc4yIdentification().template;
        delete this.ac4yIdentification.createdAt;
        delete this.ac4yIdentification.humanId;
        delete this.ac4yIdentification.id;
        return object;
    } // onlyGUIDIdentification

    noIdentification(object){
        object = object || this;
        delete this.ac4yIdentification;
        return object;
    } // noIdentification
    
    lightweight(){

        var object = this.createSelf(this);

        object = this.objectKilledForPersistence(object);

        delete object.GUID;
        delete object.createdAt;

        return object;

    } // lightweight
    
    setTemplateName(){
        this.getAc4yIdentification().getTemplate().setHumanId(this.constructor.name);
    }
    
    getGUID(){return this.getAc4yIdentification().getGUID();};
    setGUID(aGUID){return this.getAc4yIdentification().setGUID(aGUID);};
    
    getHumanId(){return this.getAc4yIdentification().getGUID();};
    setHumanId(aHumanId){return this.getAc4yIdentification().setHumanId(aHumanId);};

    check(){

        this.checkIdentification();
        this.checkIdentificationGUID();

    } // check
    
    checkIdentification() {if (!this.hasAc4yIdentification()) throw "has no identification!";}
    checkIdentificationGUID() {if (!this.getAc4yIdentification().hasGUID()) throw "has no GUID!";}

    getSerializedSize(){return new Ac4yJSONHandler().serialized(this).length;};
    
} // Ac4y

module.exports=Ac4y
