// @ac4y/ac4y-object/domain/Ac4yPersistent.js
var Ac4yPersistentAlgebra=require("@ac4y/ac4y-object/algebra/Ac4yPersistentAlgebra.js");
var Ac4yGUIDHandler = require("@ac4y/ac4y-utility/domain/Ac4yGUIDHandler.js");

var Ac4yObjectHandler=require("@ac4y/ac4y-utility/domain/Ac4yObjectHandler.js"); 

class Ac4yPersistent extends Ac4yPersistentAlgebra {

    constructor(object) {

	super(object);
        
        if (object!=undefined && object===0) return;
        
        if (!this.hasGUID()) this.setGUID(this.getAc4yIdentification().getGUID());
        /*
        if (object==undefined || object!==0) {
        //if (object) {
        
            if (!this.hasGUID()) this.setGUID(this.getAc4yIdentification().getGUID());
            
        }
        */
  /*      
        if (object!=undefined && object===0) {}
        else {
        
            if (!this.hasGUID()) this.setGUID(this.getAc4yIdentification().getGUID());
            
        }
*/
    } // constructor

    createSelf(object) {return new Ac4yPersistent(object);}

    noStandaloneIdenficationElements(object){

        object = object || this;

        delete object.createdAt;
        delete object.GUID;

        return object;

    } // noStandaloneIdenficationElements

    forPersistence(object){

        delete object.ac4yIdentification;

        //this.setCreatedAt(new Date(this.getCreatedAt()));

        if (object.hasCreatedAt()) object.setCreatedAt(new Date(object.getCreatedAt()));

        return object;

    } // forPersistence

    objectKilledForPersistence(object){

        var object = this.createSelf(object);

        object.check();

        object = new Ac4yObjectHandler().deleteObjectTypeProperties(object);
        
        return object;

    } // objectKilledForPersistence

    fromPersistence(){
        return this;
    }

    lightweight(){

        var object = this.createSelf(this);

        object = this.objectKilledForPersistence(object);

        delete object.GUID;
        delete object.createdAt;

        return object;

    } // lightweight
    
    getGUID(){
        
        if (this.hasGUID())
            return super.getGUID()
        else
        if (this.hasAc4yIdentification())
            return this.getAc4yIdentification().getGUID();
        else
            return null;
        
    } // getGUID
    
} // Ac4yPersistent

module.exports=Ac4yPersistent
