// @ac4y/ac4y-object/algebra/Ac4yIdentificationAlgebra.js
var Ac4yIdentificationBase=require("@ac4y/ac4y-object/domain/Ac4yIdentificationBase.js");

class Ac4yIdentificationAlgebra extends Ac4yIdentificationBase {

    constructor(object) {

        super(object);
        if (object != undefined && object === 0) return;
    } // constructor


    createSelf(object) {
        return new Ac4yIdentificationAlgebra(object);

    } // createSelf


    setTemplate(aTemplate) {
        this.template = aTemplate;
    }
    getTemplate() {
        return this.template;
    }
    createTemplate() {
        this.template = new Ac4yIdentificationBase();
    }
    hasTemplate() {
        return this.template != undefined;
    }


    rebuild(object, target) {

        target = target || this;
        super.rebuild(object, target);
        if (object) {

            if (object.template != undefined) target.setTemplate(new Ac4yIdentificationBase(0).rebuilded(object.template));

        } // if object does not empty

        return target;

    } // rebuild

    rebuilded(object, target) {
        return super.rebuilded(object, target);
    }

    copy(object, target) {

        target = target || this;
        super.copy(object, target);
        if (object) {

            if (object.template != undefined) target.setTemplate(object.template);

        } // if object does not empty

        return target;

    } // copy

} // Ac4yIdentificationAlgebra

module.exports=Ac4yIdentificationAlgebra
