// @ac4y/ac4y-object/algebra/Ac4yObjectAlgebra.js
var Ac4yIdentification=require("@ac4y/ac4y-object/domain/Ac4yIdentification.js");

class Ac4yObjectAlgebra {

    constructor(object) {

        if (object != undefined && object === 0) return;
        if (object) this.copy(object, this);

    } // constructor


    createSelf(object) {
        return new Ac4yObjectAlgebra(object);

    } // createSelf


    setAc4yIdentification(aAc4yIdentification) {
        this.ac4yIdentification = aAc4yIdentification;
    }
    getAc4yIdentification() {
        return this.ac4yIdentification;
    }
    createAc4yIdentification() {
        this.ac4yIdentification = new Ac4yIdentification();
    }
    hasAc4yIdentification() {
        return this.ac4yIdentification != undefined;
    }


    rebuild(object, target) {

        target = target || this;
        if (object) {

            if (object.ac4yIdentification != undefined) target.setAc4yIdentification(new Ac4yIdentification(0).rebuilded(object.ac4yIdentification));

        } // if object does not empty

        return target;

    } // rebuild

    rebuilded(object, target) {
        target = target || this;
        if (object) return this.rebuild(object, this.copy(object));

        else return target;

    } // rebuilded

    copy(object, target) {

        target = target || this;
        if (object) {

            if (object.ac4yIdentification != undefined) target.setAc4yIdentification(object.ac4yIdentification);

        } // if object does not empty

        return target;

    } // copy

} // Ac4yObjectAlgebra

module.exports=Ac4yObjectAlgebra
