// @ac4y/ac4y-object/algebra/Ac4yIdentificationBaseAlgebra.js
class Ac4yIdentificationBaseAlgebra {

    constructor(object) {

        if (object != undefined && object === 0) return;
        if (object) this.copy(object, this);

    } // constructor


    createSelf(object) {
        return new Ac4yIdentificationBaseAlgebra(object);

    } // createSelf


    setGUID(aGUID) {
        this.GUID = aGUID;
    }
    getGUID() {
        return this.GUID;
    }
    hasGUID() {
        return this.GUID != undefined;
    }

    setHumanId(aHumanId) {
        this.humanId = aHumanId;
    }
    getHumanId() {
        return this.humanId;
    }
    hasHumanId() {
        return this.humanId != undefined;
    }

    setPublicHumanId(aPublicHumanId) {
        this.publicHumanId = aPublicHumanId;
    }
    getPublicHumanId() {
        return this.publicHumanId;
    }
    hasPublicHumanId() {
        return this.publicHumanId != undefined;
    }

    setCreatedAt(aCreatedAt) {
        this.createdAt = aCreatedAt;
    }
    getCreatedAt() {
        return this.createdAt;
    }
    hasCreatedAt() {
        return this.createdAt != undefined;
    }

    setUpdatedAt(aUpdatedAt) {
        this.updatedAt = aUpdatedAt;
    }
    getUpdatedAt() {
        return this.updatedAt;
    }
    hasUpdatedAt() {
        return this.updatedAt != undefined;
    }


    rebuild(object, target) {

        target = target || this;
        if (object) {

            if (object.GUID != undefined) target.setGUID(object.GUID);
            if (object.humanId != undefined) target.setHumanId(object.humanId);
            if (object.publicHumanId != undefined) target.setPublicHumanId(object.publicHumanId);
            if (object.createdAt != undefined) target.setCreatedAt(object.createdAt);
            if (object.updatedAt != undefined) target.setUpdatedAt(object.updatedAt);

        } // if object does not empty

        return target;

    } // rebuild

    rebuilded(object, target) {
        target = target || this;
        if (object) return this.rebuild(object, this.copy(object));

        else return target;

    } // rebuilded

    copy(object, target) {

        target = target || this;
        if (object) {

            if (object.GUID != undefined) target.setGUID(object.GUID);
            if (object.humanId != undefined) target.setHumanId(object.humanId);
            if (object.publicHumanId != undefined) target.setPublicHumanId(object.publicHumanId);
            if (object.createdAt != undefined) target.setCreatedAt(object.createdAt);
            if (object.updatedAt != undefined) target.setUpdatedAt(object.updatedAt);

        } // if object does not empty

        return target;

    } // copy

} // Ac4yIdentificationBaseAlgebra

module.exports=Ac4yIdentificationBaseAlgebra
