// @ac4y/ac4y-object/algebra/Ac4yPersistentAlgebra.js
var Ac4y=require("@ac4y/ac4y-object/domain/Ac4y.js");

class Ac4yPersistentAlgebra extends Ac4y {

    constructor(object) {

        super(object);
        if (object != undefined && object === 0) return;
    } // constructor


    createSelf(object) {
        return new Ac4yPersistentAlgebra(object);

    } // createSelf


    setId(aId) {
        this.id = aId;
    }
    getId() {
        return this.id;
    }
    hasId() {
        return this.id != undefined;
    }

    setGUID(aGUID) {
        this.GUID = aGUID;
    }
    getGUID() {
        return this.GUID;
    }
    hasGUID() {
        return this.GUID != undefined;
    }

    setOwner(aOwner) {
        this.owner = aOwner;
    }
    getOwner() {
        return this.owner;
    }
    hasOwner() {
        return this.owner != undefined;
    }

    setCreatedAt(aCreatedAt) {
        this.createdAt = aCreatedAt;
    }
    getCreatedAt() {
        return this.createdAt;
    }
    hasCreatedAt() {
        return this.createdAt != undefined;
    }

    setUpdatedAt(aUpdatedAt) {
        this.updatedAt = aUpdatedAt;
    }
    getUpdatedAt() {
        return this.updatedAt;
    }
    hasUpdatedAt() {
        return this.updatedAt != undefined;
    }


    rebuild(object, target) {

        target = target || this;
        super.rebuild(object, target);
        if (object) {

            if (object.id != undefined) target.setId(object.id);
            if (object.GUID != undefined) target.setGUID(object.GUID);
            if (object.owner != undefined) target.setOwner(object.owner);
            if (object.createdAt != undefined) target.setCreatedAt(object.createdAt);
            if (object.updatedAt != undefined) target.setUpdatedAt(object.updatedAt);

        } // if object does not empty

        return target;

    } // rebuild

    rebuilded(object, target) {
        return super.rebuilded(object, target);
    }

    copy(object, target) {

        target = target || this;
        super.copy(object, target);
        if (object) {

            if (object.id != undefined) target.setId(object.id);
            if (object.GUID != undefined) target.setGUID(object.GUID);
            if (object.owner != undefined) target.setOwner(object.owner);
            if (object.createdAt != undefined) target.setCreatedAt(object.createdAt);
            if (object.updatedAt != undefined) target.setUpdatedAt(object.updatedAt);

        } // if object does not empty

        return target;

    } // copy

} // Ac4yPersistentAlgebra

module.exports=Ac4yPersistentAlgebra
