// @ac4y/ac4y-lifecycle/constant/Ac4yLifecycleConstant.js
class Ac4yLifecycleConstant {
    
    constructor() {

        this.PAUSE = "pause";
        this.LIVE = "live";
        this.STOP = "stop";
        this.ERROR = "error";
        this.BORN = "born";
        this.DEATH = "death";

        this.STATE = {
            pause: "pause"
            ,live: "live"
            ,stop: "stop"
            ,error: "error"
            ,born: "born"
            ,death: "death"
        };

        this.STAGE = {
            live: "live"
            ,idle: "idle"
            ,error: "error"
        };

        this.STEP = {
            born: "born"
            ,death: "death"
            ,error: "error"
            ,start: "start"
            ,stop: "stop"
            ,pause: "pause"
            ,continue: "continue"
        };

        this.DIMENSION = {
            state: "state"
            ,step: "step"
            ,stage: "stage"
            ,station: "station"
        };
        
    } // constructor
    
} // Ac4yLifecycleConstant

module.exports=Ac4yLifecycleConstant
