// @ac4y/ac4y-common/algebra/Ac4yBulbAlgebra.js
class Ac4yBulbAlgebra {

    constructor(object) {

        if (object != undefined && object === 0) return;
        if (object) this.copy(object, this);

    } // constructor


    createSelf(object) {
        return new Ac4yBulbAlgebra(object);

    } // createSelf


    setBulb(aBulb) {
        this.bulb = aBulb;
    }
    getBulb() {
        return this.bulb;
    }
    hasBulb() {
        return this.bulb != undefined;
    }
    isBulb() {
        return this.bulb;
    }


    rebuild(object, target) {

        target = target || this;
        if (object) {

            if (object.bulb != undefined) target.setBulb(object.bulb);

        } // if object does not empty

        return target;

    } // rebuild

    rebuilded(object, target) {
        target = target || this;
        if (object) return this.rebuild(object, this.copy(object));

        else return target;

    } // rebuilded

    copy(object, target) {

        target = target || this;
        if (object) {

            if (object.bulb != undefined) target.setBulb(object.bulb);

        } // if object does not empty

        return target;

    } // copy

} // Ac4yBulbAlgebra
		
module.exports=Ac4yBulbAlgebra
