// @ac4y/ac4y-common/algebra/Ac4yListNodeAlgebra.js
class Ac4yListNodeAlgebra {

    constructor(object) {

        if (object != undefined && object === 0) return;
        if (object) this.copy(object, this);

    } // constructor


    createSelf(object) {
        return new Ac4yListNodeAlgebra(object);

    } // createSelf


    setElement(aElement) {
        this.element = aElement;
    }
    getElement() {
        return this.element;
    }
    createElement() {
        this.element = new Array();
    }
    hasElement() {
        return this.element != undefined;
    }


    rebuild(object, target) {

        target = target || this;
        if (object) {

            if (object.element != undefined) target.setElement(object.element.slice());

        } // if object does not empty

        return target;

    } // rebuild

    rebuilded(object, target) {
        target = target || this;
        if (object) return this.rebuild(object, this.copy(object));

        else return target;

    } // rebuilded

    copy(object, target) {

        target = target || this;
        if (object) {

            if (object.element != undefined) target.setElement(object.element);

        } // if object does not empty

        return target;

    } // copy

} // Ac4yListNodeAlgebra

module.exports=Ac4yListNodeAlgebra
