// @ac4y/ac4y-timer/domain/Ac4yTimer.js
var Ac4yTimerAlgebra=require("@ac4y/ac4y-timer/algebra/Ac4yTimerAlgebra.js");
var Ac4yDateTimeHandler=require("@ac4y/ac4y-utility/domain/Ac4yDateTimeHandler.js"); 

class Ac4yTimer extends Ac4yTimerAlgebra {
     
    constructor(object){

        super(object);
        if (object != undefined && object === 0) return;
 
        if (this.hasDurationInSecond()) this.setDuration(new Ac4yDateTimeHandler().getSecondInMillisecond(this.getDurationInSecond()));
        
        if (this.hasDurationInMinute()) this.setDuration(new Ac4yDateTimeHandler().getMinuteInMillisecond(this.getDurationInMinute()));
        
    } // constructor
        
    createSelf(object) {return new Ac4yTimer(object);}
    
    getNewTimer(){return setTimeout( _ => { this.time();}, this.getDuration());};

    getRemainingTime(){
        
        if (this.isLive())
            return this.getEndTime()-new Ac4yDateTimeHandler().getSystemTime();
        else
            return 0;
    
    }; // getRemainingTime
    
    getEstimatedRunTime(){
        
        if (this.isLive())
            return this.getEndTime()-this.getStartTime();
        else
            return 0;
        
    }; // getEstimatedRunTime
    
    start(){

        super.start();

        this.setTimer(this.getNewTimer());
        
        this.setEndTime(this.getStartTime()+this.getDuration());
        
    }; // start
    
    stop(){
        
        super.stop();

        clearTimeout(this.getTimer());
        
    }; // stop
    
    cancel(){
        
        this._setStopTimeFromSystemTime();
        
        this.getDimension().state().becomesStop();
        this.getDimension().stage().reachesIdle();
        this.getDimension().step().realizesStop();
        
        this.getLiveIndicator().switchOff();
        
        this._tryFireOnCancel();
        
        this._destroy();        

        clearTimeout(this.getTimer());
        
    }; // stop
    
    time(){
        
        this.getLiveIndicator().switchOff();
        
        this._setStopTimeFromSystemTime();

        this._tryFireOnTime();
        
//        this.stop();
        
    }; // time
    
    getInformationObject(){
        
        return {
            "név": this.getName()
            ,"ütemezett idő": this.getDuration()
            ,"ütemezett idő (mp)": new Ac4yDateTimeHandler().getMillisecondInSecond(this.getDuration())
            ,"elindult": new Ac4yDateTimeHandler().getDateTimeExternalFormat(new Date(this.getStartTime()))
            ,"befezés": new Ac4yDateTimeHandler().getDateTimeExternalFormat(new Date(this.getEndTime()))
            ,"leállt": new Ac4yDateTimeHandler().getDateTimeExternalFormat(new Date(this.getStopTime()))
            ,"hátralévő idő": this.getRemainingTime()
            ,"hátralévő idő (mp)": new Ac4yDateTimeHandler().getMillisecondInSecond(this.getRemainingTime())
            ,"dolgozik": this.isLive()
        };
        
    } // getInformationObject
    
    getShortInformationObject(){
        
        return {
            "időzítő neve" : this.getName()
            ,"hátralévő idő (mp)" : 
               this.isLive() ? new Ac4yDateTimeHandler().getMillisecondInSecond(
                                        this.getRemainingTime()
                                ) : "nem üzemel"
        };
        
    } // getShortInformationObject
    
} // Ac4yTimer

module.exports=Ac4yTimer
