// @ac4y/ac4y-common/state/Ac4yStateSetList.js
var Ac4yListNode=require("@ac4y/ac4y-common/container/Ac4yListNode.js");
var Ac4yStateSet=require("@ac4y/ac4y-common/state/Ac4yStateSet.js");
var Ac4yDateTimeHandler=require("@ac4y/ac4y-common/utility/Ac4yDateTimeHandler.js");

class Ac4yStateSetList extends Ac4yListNode {
    
    addStateSet(state){
        this.addElement(
            new Ac4yStateSet({
                state: state
                ,timestamp: new Ac4yDateTimeHandler().getSystemTime()
            })
        );
    }
    
    getStateFiltered(aState){
        
        return this.get().filter(
                        function (aStateSet) {
                            return (aStateSet.getState()===aState)
                        }                    
                    );
            
    } // getFiltered
    
    getSortedByState(){
        
        return this.get().sort(
                        function(aStateSetOne, aStateSetOtherOne){
                            return aStateSetOne.getState()>aStateSetOtherOne.getState()
                        }
                ); 
        
    } // getSortedByState

    getReverseSortedByState(){
        
        return this.get().sort(
                        function(aStateSetOne, aStateSetOtherOne){
                            return aStateSetOne.getState()<aStateSetOtherOne.getState()
                        }
                ); 
        
    } // getReverseSortedByState

    getSortedByTimestamp(){
        
        return this.get().sort(
                        function(aStateSetOne, aStateSetOtherOne){
                            return aStateSetOne.getTimestamp()-aStateSetOtherOne.getTimestamp()
                        }
                ); 
        
    } // getSortedByTimestamp

    getReverseSortedByTimestamp(){
        
        return this.get().sort(
                        function(aStateSetOne, aStateSetOtherOne){
                            return aStateSetOtherOne.getTimestamp()-aStateSetOne.getTimestamp()
                        }
                ); 
        
    } // getReverseSortedByTimestamp
    
    getTheYoungest(){return new Ac4yStateSetList(this.getSortedByTimestamp()).getFirst();}
    
    getTheOldest(){return new Ac4yStateSetList(this.getSortedByTimestamp()).getLast();}    
   
    hasOccurenceOfThisState(aState){return !new Ac4yStateSetList(this.getStateFiltered(aState)).isEmpty();}
    
    hasReachedThisState(aState){return this.hasOccurenceOfThisState(aState);}
    
    getTheYoungestOfThisState(aState){
        
        return new Ac4yStateSetList(
                    new Ac4yStateSetList(
                        this.getStateFiltered(aState)
                    ).getSortedByTimestamp()
               ).getLast();
        
    } // getTheYoungestOfThisState
    
    getTheOldestOfThisState(aState){
        
        return new Ac4yStateSetList(
                    new Ac4yStateSetList(
                        this.getStateFiltered(aState)
                    ).getSortedByTimestamp()
               ).getFirst();
        
    } // getTheOldestOfThisState
    
    getTimestampOfTheYoungestOfThisState(aState){return this.getTheYoungestOfThisState(aState).getTimestamp();}

    getTimestampOfTheOldestOfThisState(aState){return this.getTheOldestOfThisState(aState).getTimestamp();}
    
} // Ac4yStateSetList

module.exports=Ac4yStateSetList
