// @ac4y/ac4y-service/algebra/Ac4yServiceEnvelopeAlgebra.js
class Ac4yServiceEnvelopeAlgebra {

    constructor(object) {

        if (object != undefined && object === 0) return;
        if (object) this.copy(object, this);

    } // constructor


    createSelf(object) {
        return new Ac4yServiceEnvelopeAlgebra(object);

    } // createSelf


    setSession(aSession) {
        this.session = aSession;
    }
    getSession() {
        return this.session;
    }
    hasSession() {
        return this.session != undefined;
    }

    setOwner(aOwner) {
        this.owner = aOwner;
    }
    getOwner() {
        return this.owner;
    }
    hasOwner() {
        return this.owner != undefined;
    }


    rebuild(object, target) {

        target = target || this;
        if (object) {

            if (object.session != undefined) target.setSession(object.session);
            if (object.owner != undefined) target.setOwner(object.owner);

        } // if object does not empty

        return target;

    } // rebuild

    rebuilded(object, target) {
        target = target || this;
        if (object) return this.copy(object, this.rebuild(object));

        else return target;

    } // rebuilded

    copy(object, target) {

        target = target || this;
        if (object) {

            if (object.session != undefined) target.setSession(object.session);
            if (object.owner != undefined) target.setOwner(object.owner);

        } // if object does not empty

        return target;

    } // copy

} // Ac4yServiceEnvelopeAlgebra
		
module.exports=Ac4yServiceEnvelopeAlgebra
