// @ac4y/ac4y-service/algebra/Ac4yServiceAlgebra.js
var Ac4yFlow=require("@ac4y/ac4y-flow/domain/Ac4yFlow.js"); 
var Ac4ySLCDStock=require("../domain/Ac4ySLCDStock.js");
var Ac4yServiceRequest=require("../domain/Ac4yServiceRequest.js");
var Ac4yServiceResponse=require("../domain/Ac4yServiceResponse.js");
var Ac4yKeyValueMemory=require("@ac4y/ac4y-common/container/Ac4yKeyValueMemory.js");

class Ac4yServiceAlgebra extends Ac4yFlow {

    constructor(object) {

        super(object);
        if (object != undefined && object === 0) return;
    } // constructor


    createSelf(object) {
        return new Ac4yServiceAlgebra(object);

    } // createSelf


    setDimension(aDimension) {
        this.dimension = aDimension;
    }
    getDimension() {
        return this.dimension;
    }
    createDimension() {
        this.dimension = new Ac4ySLCDStock();
    }
    hasDimension() {
        return this.dimension != undefined;
    }

    setDataSafe(aDataSafe) {
        this.dataSafe = aDataSafe;
    }
    getDataSafe() {
        return this.dataSafe;
    }
    hasDataSafe() {
        return this.dataSafe != undefined;
    }

    setOnCommit(aOnCommit) {
        this.onCommit = aOnCommit;
    }
    getOnCommit() {
        return this.onCommit;
    }
    hasOnCommit() {
        return this.onCommit != undefined;
    }
    _fireOnCommit() {
        this.onCommit();
    };
    _tryFireOnCommit() {
        if (this.onCommit) this._fireOnCommit();
    };

    setOnRollback(aOnRollback) {
        this.onRollback = aOnRollback;
    }
    getOnRollback() {
        return this.onRollback;
    }
    hasOnRollback() {
        return this.onRollback != undefined;
    }
    _fireOnRollback() {
        this.onRollback();
    };
    _tryFireOnRollback() {
        if (this.onRollback) this._fireOnRollback();
    };

    setOnSuccess(aOnSuccess) {
        this.onSuccess = aOnSuccess;
    }
    getOnSuccess() {
        return this.onSuccess;
    }
    hasOnSuccess() {
        return this.onSuccess != undefined;
    }
    _fireOnSuccess() {
        this.onSuccess();
    };
    _tryFireOnSuccess() {
        if (this.onSuccess) this._fireOnSuccess();
    };

    setOnFail(aOnFail) {
        this.onFail = aOnFail;
    }
    getOnFail() {
        return this.onFail;
    }
    hasOnFail() {
        return this.onFail != undefined;
    }
    _fireOnFail() {
        this.onFail();
    };
    _tryFireOnFail() {
        if (this.onFail) this._fireOnFail();
    };

    setOnNothingHappened(aOnNothingHappened) {
        this.onNothingHappened = aOnNothingHappened;
    }
    getOnNothingHappened() {
        return this.onNothingHappened;
    }
    hasOnNothingHappened() {
        return this.onNothingHappened != undefined;
    }
    _fireOnNothingHappened() {
        this.onNothingHappened();
    };
    _tryFireOnNothingHappened() {
        if (this.onNothingHappened) this._fireOnNothingHappened();
    };

    setOnDone(aOnDone) {
        this.onDone = aOnDone;
    }
    getOnDone() {
        return this.onDone;
    }
    hasOnDone() {
        return this.onDone != undefined;
    }
    _fireOnDone() {
        this.onDone();
    };
    _tryFireOnDone() {
        if (this.onDone) this._fireOnDone();
    };

    setRequest(aRequest) {
        this.request = aRequest;
    }
    getRequest() {
        return this.request;
    }
    createRequest() {
        this.request = new Ac4yServiceRequest();
    }
    hasRequest() {
        return this.request != undefined;
    }

    setResponse(aResponse) {
        this.response = aResponse;
    }
    getResponse() {
        return this.response;
    }
    createResponse() {
        this.response = new Ac4yServiceResponse();
    }
    hasResponse() {
        return this.response != undefined;
    }


    rebuild(object, target) {

        target = target || this;
        super.rebuild(object, target);
        if (object) {

            if (object.dimension != undefined) target.setDimension(new Ac4ySLCDStock(0).rebuilded(object.dimension));
            if (object.dataSafe != undefined) target.setDataSafe(object.dataSafe);
            if (object.onCommit != undefined) target.setOnCommit(object.onCommit);
            if (object.onRollback != undefined) target.setOnRollback(object.onRollback);
            if (object.onSuccess != undefined) target.setOnSuccess(object.onSuccess);
            if (object.onFail != undefined) target.setOnFail(object.onFail);
            if (object.onNothingHappened != undefined) target.setOnNothingHappened(object.onNothingHappened);
            if (object.onDone != undefined) target.setOnDone(object.onDone);
            if (object.request != undefined) target.setRequest(new Ac4yServiceRequest(0).rebuilded(object.request));
            if (object.response != undefined) target.setResponse(new Ac4yServiceResponse(0).rebuilded(object.response));

        } // if object does not empty

        return target;

    } // rebuild

    rebuilded(object, target) {
        return super.rebuilded(object, target);
    }

    copy(object, target) {

        target = target || this;
        super.copy(object, target);
        if (object) {

            if (object.dimension != undefined) target.setDimension(object.dimension);
            if (object.dataSafe != undefined) target.setDataSafe(object.dataSafe);
            if (object.onCommit != undefined) target.setOnCommit(object.onCommit);
            if (object.onRollback != undefined) target.setOnRollback(object.onRollback);
            if (object.onSuccess != undefined) target.setOnSuccess(object.onSuccess);
            if (object.onFail != undefined) target.setOnFail(object.onFail);
            if (object.onNothingHappened != undefined) target.setOnNothingHappened(object.onNothingHappened);
            if (object.onDone != undefined) target.setOnDone(object.onDone);
            if (object.request != undefined) target.setRequest(object.request);
            if (object.response != undefined) target.setResponse(object.response);

        } // if object does not empty

        return target;

    } // copy

} // Ac4yServiceAlgebra
		
module.exports=Ac4yServiceAlgebra
