// @ac4y/ac4y-service/algebra/Ac4yProcessResultAlgebra.js
class Ac4yProcessResultAlgebra {

    constructor(object) {

        if (object != undefined && object === 0) return;
        if (object) this.copy(object, this);

    } // constructor


    createSelf(object) {
        return new Ac4yProcessResultAlgebra(object);

    } // createSelf


    setRequestId(aRequestId) {
        this.requestId = aRequestId;
    }
    getRequestId() {
        return this.requestId;
    }
    hasRequestId() {
        return this.requestId != undefined;
    }

    setCode(aCode) {
        this.code = aCode;
    }
    getCode() {
        return this.code;
    }
    hasCode() {
        return this.code != undefined;
    }

    setMessage(aMessage) {
        this.message = aMessage;
    }
    getMessage() {
        return this.message;
    }
    hasMessage() {
        return this.message != undefined;
    }

    setDescription(aDescription) {
        this.description = aDescription;
    }
    getDescription() {
        return this.description;
    }
    hasDescription() {
        return this.description != undefined;
    }


    rebuild(object, target) {

        target = target || this;
        if (object) {

            if (object.requestId != undefined) target.setRequestId(object.requestId);
            if (object.code != undefined) target.setCode(object.code);
            if (object.message != undefined) target.setMessage(object.message);
            if (object.description != undefined) target.setDescription(object.description);

        } // if object does not empty

        return target;

    } // rebuild

    rebuilded(object, target) {
        target = target || this;
        if (object) return this.copy(object, this.rebuild(object));

        else return target;

    } // rebuilded

    copy(object, target) {

        target = target || this;
        if (object) {

            if (object.requestId != undefined) target.setRequestId(object.requestId);
            if (object.code != undefined) target.setCode(object.code);
            if (object.message != undefined) target.setMessage(object.message);
            if (object.description != undefined) target.setDescription(object.description);

        } // if object does not empty

        return target;

    } // copy

} // Ac4yProcessResultAlgebra
		
module.exports=Ac4yProcessResultAlgebra
