// @ac4y/ac4y-service/algebra/Ac4yServiceRequestAlgebra.js
var Ac4yServiceEnvelope=require("../domain/Ac4yServiceEnvelope.js");
var Ac4y=require("@ac4y/ac4y-object/domain/Ac4y.js"); 

class Ac4yServiceRequestAlgebra extends Ac4y {

    constructor(object) {

        super(object);
        if (object != undefined && object === 0) return;
    } // constructor


    createSelf(object) {
        return new Ac4yServiceRequestAlgebra(object);

    } // createSelf


    setEnvelope(aEnvelope) {
        this.envelope = aEnvelope;
    }
    getEnvelope() {
        return this.envelope;
    }
    createEnvelope() {
        this.envelope = new Ac4yServiceEnvelope();
    }
    hasEnvelope() {
        return this.envelope != undefined;
    }


    rebuild(object, target) {

        target = target || this;
        super.rebuild(object, target);
        if (object) {

            if (object.envelope != undefined) target.setEnvelope(new Ac4yServiceEnvelope(0).rebuilded(object.envelope));

        } // if object does not empty

        return target;

    } // rebuild

    rebuilded(object, target) {
        return super.rebuilded(object, target);
    }

    copy(object, target) {

        target = target || this;
        super.copy(object, target);
        if (object) {

            if (object.envelope != undefined) target.setEnvelope(object.envelope);

        } // if object does not empty

        return target;

    } // copy

} // Ac4yServiceRequestAlgebra
		
module.exports=Ac4yServiceRequestAlgebra
