// @ac4y/ac4y-service/algebra/Ac4yServiceResponseAlgebra.js
var Ac4y=require("@ac4y/ac4y-object/domain/Ac4y.js"); 
var Ac4yProcessResult=require("../domain/Ac4yProcessResult.js");
var Ac4ySLCDStock=require("../domain/Ac4ySLCDStock.js");
var Ac4yKeyValueMemory=require("@ac4y/ac4y-common/container/Ac4yKeyValueMemory.js");

class Ac4yServiceResponseAlgebra extends Ac4y {

    constructor(object) {

        super(object);
        if (object != undefined && object === 0) return;
    } // constructor


    createSelf(object) {
        return new Ac4yServiceResponseAlgebra(object);

    } // createSelf


    setResult(aResult) {
        this.result = aResult;
    }
    getResult() {
        return this.result;
    }
    createResult() {
        this.result = new Ac4yProcessResult();
    }
    hasResult() {
        return this.result != undefined;
    }

    setDimension(aDimension) {
        this.dimension = aDimension;
    }
    getDimension() {
        return this.dimension;
    }
    createDimension() {
        this.dimension = new Ac4ySLCDStock();
    }
    hasDimension() {
        return this.dimension != undefined;
    }

    setIndicatorSet(aIndicatorSet) {
        this.indicatorSet = aIndicatorSet;
    }
    getIndicatorSet() {
        return this.indicatorSet;
    }
    createIndicatorSet() {
        this.indicatorSet = new Ac4yKeyValueMemory();
    }
    hasIndicatorSet() {
        return this.indicatorSet != undefined;
    }

    setSubset(aSubset) {
        this.subset = aSubset;
    }
    getSubset() {
        return this.subset;
    }
    createSubset() {
        this.subset = new Ac4yArray();
    }
    hasSubset() {
        return this.subset != undefined;
    }

    setThreadId(aThreadId) {
        this.threadId = aThreadId;
    }
    getThreadId() {
        return this.threadId;
    }
    hasThreadId() {
        return this.threadId != undefined;
    }


    rebuild(object, target) {

        target = target || this;
        super.rebuild(object, target);
        if (object) {

            if (object.result != undefined) target.setResult(new Ac4yProcessResult(0).rebuilded(object.result));
            if (object.dimension != undefined) target.setDimension(new Ac4ySLCDStock(0).rebuilded(object.dimension));
            if (object.indicatorSet != undefined) target.setIndicatorSet(new Ac4yKeyValueMemory(0).rebuilded(object.indicatorSet));
            if (object.subset != undefined) target.setSubset(new Ac4yArray(0).rebuilded(object.subset));
            if (object.threadId != undefined) target.setThreadId(object.threadId);

        } // if object does not empty

        return target;

    } // rebuild

    rebuilded(object, target) {
        return super.rebuilded(object, target);
    }

    copy(object, target) {

        target = target || this;
        super.copy(object, target);
        if (object) {

            if (object.result != undefined) target.setResult(object.result);
            if (object.dimension != undefined) target.setDimension(object.dimension);
            if (object.indicatorSet != undefined) target.setIndicatorSet(object.indicatorSet);
            if (object.subset != undefined) target.setSubset(object.subset);
            if (object.threadId != undefined) target.setThreadId(object.threadId);

        } // if object does not empty

        return target;

    } // copy

} // Ac4yServiceResponseAlgebra
		
module.exports=Ac4yServiceResponseAlgebra
