// @ac4y/ac4y-service/algebra/Ac4yStatelessServiceAlgebra.js
var Ac4y=require("@ac4y/ac4y-object/domain/Ac4y.js"); 

class Ac4yStatelessServiceAlgebra extends Ac4y {

    constructor(object) {

        super(object);
        if (object != undefined && object === 0) return;
    } // constructor


    createSelf(object) {
        return new Ac4yStatelessServiceAlgebra(object);

    } // createSelf


    setOnStart(aOnStart) {
        this.onStart = aOnStart;
    }
    getOnStart() {
        return this.onStart;
    }
    hasOnStart() {
        return this.onStart != undefined;
    }
    _fireOnStart() {
        this.onStart();
    };
    _tryFireOnStart() {
        if (this.onStart) this._fireOnStart();
    };

    setOnRequest(aOnRequest) {
        this.onRequest = aOnRequest;
    }
    getOnRequest() {
        return this.onRequest;
    }
    hasOnRequest() {
        return this.onRequest != undefined;
    }
    _fireOnRequest() {
        this.onRequest();
    };
    _tryFireOnRequest() {
        if (this.onRequest) this._fireOnRequest();
    };

    setOnCommit(aOnCommit) {
        this.onCommit = aOnCommit;
    }
    getOnCommit() {
        return this.onCommit;
    }
    hasOnCommit() {
        return this.onCommit != undefined;
    }
    _fireOnCommit() {
        this.onCommit();
    };
    _tryFireOnCommit() {
        if (this.onCommit) this._fireOnCommit();
    };

    setOnRollback(aOnRollback) {
        this.onRollback = aOnRollback;
    }
    getOnRollback() {
        return this.onRollback;
    }
    hasOnRollback() {
        return this.onRollback != undefined;
    }
    _fireOnRollback() {
        this.onRollback();
    };
    _tryFireOnRollback() {
        if (this.onRollback) this._fireOnRollback();
    };


    rebuild(object, target) {

        target = target || this;
        super.rebuild(object, target);
        if (object) {

            if (object.onStart != undefined) target.setOnStart(object.onStart);
            if (object.onRequest != undefined) target.setOnRequest(object.onRequest);
            if (object.onCommit != undefined) target.setOnCommit(object.onCommit);
            if (object.onRollback != undefined) target.setOnRollback(object.onRollback);

        } // if object does not empty

        return target;

    } // rebuild

    rebuilded(object, target) {
        return super.rebuilded(object, target);
    }

    copy(object, target) {

        target = target || this;
        super.copy(object, target);
        if (object) {

            if (object.onStart != undefined) target.setOnStart(object.onStart);
            if (object.onRequest != undefined) target.setOnRequest(object.onRequest);
            if (object.onCommit != undefined) target.setOnCommit(object.onCommit);
            if (object.onRollback != undefined) target.setOnRollback(object.onRollback);

        } // if object does not empty

        return target;

    } // copy

} // Ac4yStatelessServiceAlgebra

module.exports=Ac4yStatelessServiceAlgebra
