// @ac4y/ac4y-service/algebra/GetByValueRequestAlgebra.js
class GetByValueRequestAlgebra {

    constructor(aObject) {

        if (aObject) this.rebuild(aObject, this);

    } // constructor


    createSelf() {
        return new GetByValueRequestAlgebra();

    } // createSelf

    setValue(aValue) {
        this.value = aValue;
    }
    getValue() {
        return this.value;
    }
    hasValue() {
        return this.value != undefined;
    }


    rebuild(aObject, aTarget) {

        if (aObject) {

            if (aObject.value != undefined) aTarget.setValue(aObject.value);

        } // if aObject does not empty

        return aTarget;

    } // rebuild

    getNew(aValue) {

        var vObject = this.createSelf();

        if (aValue) vObject.setValue(aValue);

        return vObject;

    } // getNew

    getNewFromObject(aObject) {
        return this.rebuild(aObject, this.createSelf());
    }

    serialized() {
        return JSON.stringify(JSON.decycle(this));
    }

    deserialized(aSerialized) {
        return this.getNewFromObject(JSON.parse(aSerialized));
    }

    hibernate() {
        new LocalStorageHandler().put('GetByValueRequest', new Ac4yBase64Handler().encode(this.serialized()));
    };

    rehibernate() {
        this.rebuild(this.deserialized(new Ac4yBase64Handler().decode(new LocalStorageHandler().get('GetByValueRequest'))), this);
    };

    getNewFromHibernation() {
        var vObject = this.createSelf();
        vObject.rehibernate();
        return vObject;
    };

} // GetByValueRequestAlgebra

module.exports=GetByValueRequestAlgebra
