// @ac4y/ac4y-service/algebra/GetByIdRequestAlgebra.js
class GetByIdRequestAlgebra {

    constructor(aObject) {

        if (aObject) this.rebuild(aObject, this);

    } // constructor


    createSelf() {
        return new GetByIdRequestAlgebra();

    } // createSelf

    setId(aId) {
        this.id = aId;
    }
    getId() {
        return this.id;
    }
    hasId() {
        return this.id != undefined;
    }


    rebuild(aObject, aTarget) {

        if (aObject) {

            if (aObject.id != undefined) aTarget.setId(aObject.id);

        } // if aObject does not empty

        return aTarget;

    } // rebuild

    getNew(aId) {

        var vObject = this.createSelf();

        if (aId) vObject.setId(aId);

        return vObject;

    } // getNew

    getNewFromObject(aObject) {
        return this.rebuild(aObject, this.createSelf());
    }

    serialized() {
        return JSON.stringify(JSON.decycle(this));
    }

    deserialized(aSerialized) {
        return this.getNewFromObject(JSON.parse(aSerialized));
    }

    hibernate() {
        new LocalStorageHandler().put('GetByIdRequest', new Ac4yBase64Handler().encode(this.serialized()));
    };

    rehibernate() {
        this.rebuild(this.deserialized(new Ac4yBase64Handler().decode(new LocalStorageHandler().get('GetByIdRequest'))), this);
    };

    getNewFromHibernation() {
        var vObject = this.createSelf();
        vObject.rehibernate();
        return vObject;
    };

} // GetByIdRequestAlgebra

module.exports=GetByIdRequestAlgebra
