// @ac4y/ac4y-service/algebra/GetListResponseAlgebra.js
var Ac4yProcessResult=require("../domain/Ac4yProcessResult.js"); 

class GetListResponseAlgebra {

    constructor(aObject) {

        if (aObject) this.rebuild(aObject, this);

    } // constructor


    createSelf() {
        return new GetListResponseAlgebra();

    } // createSelf

    setResult(aResult) {
        this.result = aResult;
    }
    getResult() {
        return this.result;
    }
    createResult() {
        this.result = new Ac4yProcessResult();
    }
    hasResult() {
        return this.result != undefined;
    }

    setList(aList) {
        this.list = aList;
    }
    getList() {
        return this.list;
    }
    createList() {
        this.list = new Object();
    }
    hasList() {
        return this.list != undefined;
    }


    rebuild(aObject, aTarget) {

        if (aObject) {

            if (aObject.result != undefined) aTarget.setResult(new Ac4yProcessResult(aObject.result));
            if (aObject.list != undefined) aTarget.setList(new Object(aObject.list));

        } // if aObject does not empty

        return aTarget;

    } // rebuild

    getNew(aResult, aList) {

        var vObject = this.createSelf();

        if (aResult) vObject.setResult(aResult);
        if (aList) vObject.setList(aList);

        return vObject;

    } // getNew

    getNewFromObject(aObject) {
        return this.rebuild(aObject, this.createSelf());
    }

    serialized() {
        return JSON.stringify(JSON.decycle(this));
    }

    deserialized(aSerialized) {
        return this.getNewFromObject(JSON.parse(aSerialized));
    }

    hibernate() {
        new LocalStorageHandler().put('GetListResponse', new Ac4yBase64Handler().encode(this.serialized()));
    };

    rehibernate() {
        this.rebuild(this.deserialized(new Ac4yBase64Handler().decode(new LocalStorageHandler().get('GetListResponse'))), this);
    };

    getNewFromHibernation() {
        var vObject = this.createSelf();
        vObject.rehibernate();
        return vObject;
    };

} // GetListResponseAlgebra

module.exports=GetListResponseAlgebra
