// @ac4y/ac4y-service/algebra/GetObjectResponseAlgebra.js
var Ac4yProcessResult=require("../domain/Ac4yProcessResult.js"); 
var Ac4yServiceResponse=require("../domain/Ac4yServiceResponse.js"); 

class GetObjectResponseAlgebra extends Ac4yServiceResponse {

    constructor(object) {

        super(object);
        if (object != undefined && object === 0) return;
    } // constructor


    createSelf(object) {
        return new GetObjectResponseAlgebra(object);

    } // createSelf


    setObject(aObject) {
        this.object = aObject;
    }
    getObject() {
        return this.object;
    }
    createObject() {
        this.object = new Object();
    }
    hasObject() {
        return this.object != undefined;
    }


    rebuild(object, target) {

        target = target || this;
        super.rebuild(object, target);
        if (object) {

            if (object.object != undefined) target.setObject(new Object(0).rebuilded(object.object));

        } // if object does not empty

        return target;

    } // rebuild

    rebuilded(object, target) {
        return super.rebuilded(object, target);
    }

    copy(object, target) {

        target = target || this;
        super.copy(object, target);
        if (object) {

            if (object.object != undefined) target.setObject(object.object);

        } // if object does not empty

        return target;

    } // copy

} // GetObjectResponseAlgebra

module.exports=GetObjectResponseAlgebra
