// @ac4y/ac4y-utility/domain/Ac4yDateTimeHandler.js
var Ac4yStringHandler=require("./Ac4yStringHandler.js"); 

class Ac4yDateTimeHandler {
    
    getSystemTime(){return new Date().getTime();}
  
    getSystemTimeInExternalFormat(){return this.getSystemTime();}
    
    getDateExternalFormat(aDate) {
        
      var vYear = aDate.getFullYear();
      var vMonth = aDate.getMonth()+1;
      var vDay = aDate.getDate();
      
      return new Ac4yStringHandler().padNumber(vYear,4)+"-"
              +new Ac4yStringHandler().padNumber(vMonth,2)+"-"
              +new Ac4yStringHandler().padNumber(vDay,2);
      
    } // getTimeExternalFormat
    
    getTimeExternalFormat(aDatetime) {
        
      var vHour = aDatetime.getHours();
      var vMinute = aDatetime.getMinutes();
      var vSecond = aDatetime.getSeconds();
      
      return new Ac4yStringHandler().padNumber(vHour,2)
              +":"+new Ac4yStringHandler().padNumber(vMinute,2)
              +":"+new Ac4yStringHandler().padNumber(vSecond,2);
      
    } // getTimeExternalFormat
    
    getDateTimeExternalFormat(aDatetime) {
        
        return aDatetime.toJSON();
      
    } // getDateTimeExternalFormat
    
    
    shiftDay(date, shift){

        return new Date(new Date(date).setDate(new Date(date).getDate() + shift ));

    } // shiftDay

    getMillisecondVsSecond(){return 1000;}
    getSecondVsMinute(){return 60;}
    getMinuteVsHour(){return 60;}
    getHourVsDay(){return 24;}
    
    getMillisecondInSecond(aMillisecond){return Math.round(aMillisecond / this.getMillisecondVsSecond());}
    getMillisecondInMinute(aMillisecond){return Math.round(this.getMillisecondInSecond(aMillisecond) / this.getSecondVsMinute());}
    getMillisecondInHour(aMillisecond){return Math.round(this.getMillisecondInMinute(aMillisecond) / this.getMinuteVsHour());}
    getMillisecondInDay(aMillisecond){return Math.round(this.getMillisecondInHour(aMillisecond) / this.getHourVsDay());}

    //getSecondInMillisecond(aSecond){return aSecond * this.getMillisecondVsSecond();}
    getMinuteInMillisecond(aMinute){return this.getSecondInMillisecond(aSecond) * this.getSecondVsMinute();}
    getHourInMillisecond(aHour){return this.getSecondInMinute(aSecond) * this.getMinuteVsHour();}
    getDayInMillisecond(aDay){return this.getSecondInHour(aSecond) * this.getHourVsDay();}
    
    getSecondInMillisecond(aSecond){return aSecond * this.getMillisecondVsSecond();}
    getSecondInMinute(aSecond){return this.getSecondInMillisecond(aSecond) * this.getSecondVsMinute();}
    getSecondInHour(aSecond){return this.getSecondInMinute(aSecond) * this.getMinuteVsHour();}
    getSecondInDay(aSecond){return this.getSecondInHour(aSecond) * this.getHourVsDay();}
    /*
    getMinuteInMillisecond(aMinute){return aMinute * this.getMillisecondVsSecond();}
    getMinuteInSecond(aMinute){return this.getMinuteInMillisecond(aMinute) * this.getSecondVsMinute();}
    getMinuteInHour(aMinute){return this.getMinuteInSecond(aMinute) * this.getMinuteVsHour();}
    getMinuteInDay(aMinute){return this.getMinuteInHour(aMinute) * this.getHourVsDay();}
    */
   
    sleep(milliseconds) {
        var start = new Date().getTime();
        for (var i = 0; i < 1e7; i++) {
            if ((new Date().getTime() - start) > milliseconds){
                break;
            }
        }
    } // sleep
    
} // Ac4yDateTimeHandler

module.exports=Ac4yDateTimeHandler
