// @ac4y/ac4y-utility/domain/Ac4yObjectHandler.js
class Ac4yObjectHandler {

    getObjectPropertiesAsVerticalTable(aObject, aDocument, aOwner){
        
        var vTable = new Ac4yDOMHandler().getCreatedTable(aDocument, aOwner);

        if (aObject)
            Object.getOwnPropertyNames(aObject).forEach(

                function (aValue, aIndex, aArray) {

                    var vRow = vTable.insertRow(-1);

                    vRow.insertCell(-1).innerHTML = aValue;
                    vRow.insertCell(-1).innerHTML = aObject[aValue];

                }

            );

        return vTable;
        
    } // getObjectPropertiesAsVerticalTable

    getObjectPropertiesAsHorizontalTable(aObject, aDocument, aOwner){
        
        var vTable = new Ac4yDOMHandler().getCreatedTable(aDocument, aOwner);

        var vRow = vTable.insertRow(-1);
        
        if (aObject)
            Object.getOwnPropertyNames(aObject).forEach(
                function (aValue, aIndex, aArray) {

                    vRow.insertCell(-1).innerHTML = aValue;

                }

            );

        var vRow = vTable.insertRow(-1);

        if (aObject)
            Object.getOwnPropertyNames(aObject).forEach(

                function (aValue, aIndex, aArray) {

                    vRow.insertCell(-1).innerHTML = aObject[aValue];

                }

            );

        return vTable;
        
    } // getObjectPropertiesAsHorizontalTable
    
    getArrayAsVerticalTable(aObject, aDocument, aOwner){
        
        var vTable = null;
        
        if (aObject)
            if (aObject.length)
                if (aObject.length>0) {

                    vTable = new Ac4yDOMHandler().getCreatedTable(aDocument, aOwner);

                    var vRow = vTable.insertRow(-1);

                    if (aObject && aObject[0])
                        Object.getOwnPropertyNames(aObject[0]).forEach(
                            function (aKey, aIndex, aArray) {
                                vRow.insertCell(-1).innerHTML = aKey;
                            }
                        );

                    for (var vIndex = 0; vIndex < aObject.length; vIndex++) { 
                        
                        var vRow = vTable.insertRow(-1);

                        if (aObject && aObject[vIndex])
                            Object.getOwnPropertyNames(aObject[vIndex]).forEach(
                                function (aKey, aIndex, aArray) {
                                    vRow.insertCell(-1).innerHTML = aObject[vIndex][aKey];
                                }
                            );
                    
                    }
                    
        } // if
        
        return vTable;
        
    } // getArrayAsVerticalTable

    getArrayAsHorizontalTable(aObject, aDocument, aOwner){

        var vTable = null;

        if (
                aObject 
                && aObject.length 
                && aObject.length>0 
                && Object.getOwnPropertyNames(aObject[0]) 
                && Object.getOwnPropertyNames(aObject[0]).length
            ) {
        
                var vRowCount = Object.getOwnPropertyNames(aObject[0]).length;
                var vColumnCount = aObject.length;

                vTable = new Ac4yDOMHandler().getCreatedTable(aDocument, aOwner);

                var vRow = vTable.insertRow(-1);

                for (var vRowIndex = 0; vRowIndex < vRowCount; vRowIndex++) {

                    var vRow = vTable.insertRow(-1);

                    vRow.insertCell(-1).innerHTML = Object.getOwnPropertyNames(aObject[0])[vRowIndex];

                    for (var vColumnIndex = 0; vColumnIndex < vColumnCount; vColumnIndex++) {

                        vRow.insertCell(-1).innerHTML = aObject[vColumnIndex][(Object.getOwnPropertyNames(aObject[0])[vRowIndex])];

                    } // for columns

                } // for rows
                
        } // if

        return vTable;

    } // getArrayAsHorizontalTable

    doesExist(aObject){return (aObject ? true : false);}
    
    isThisNothing(aObject){return (aObject ? false : true);}
    
    isCorrect(aObject){return (aObject ? true : false);}
    
    isFunction(aObject){return (((aObject) && (typeof aObject === 'function')) ? true : false );}
    
    default(aValue, aDefaultValue){

        if (!this.isThisNothing(aValue))            
            return aValue;
        else
            return aDefaultValue;
            
    } // default

    getClassNameOfTheObject(aObject){return aObject.constructor.name;};

    deleteObjectTypeProperties(object){

        if (!object) throw "object is null!";
        Object.getOwnPropertyNames(Object.getOwnPropertyDescriptors(object)).forEach( (key, index) => {

            var property=Object.getOwnPropertyDescriptors(object)[key];

            if (typeof property.value === "object" || typeof property.value === "function")
                delete object[key];

        });

        return object;

    } // deleteObjectTypeProperties 
    
} // Ac4yObjectHandler

module.exports=Ac4yObjectHandler
