// @ac4y/ac4y-utility/domain/Ac4yJSONHandler.js
var cycle=require("./cycle.js");

/*
exports = module.exports = stringify
exports.getSerialize = serializer

function stringify(obj, replacer, spaces, cycleReplacer) {
  return JSON.stringify(obj, serializer(replacer, cycleReplacer), spaces)
}

function serializer(replacer, cycleReplacer) {
  var stack = [], keys = []

  if (cycleReplacer == null) cycleReplacer = function(key, value) {
    if (stack[0] === value) return "[Circular ~]"
    return "[Circular ~." + keys.slice(0, stack.indexOf(value)).join(".") + "]"
  }

  return function(key, value) {
    if (stack.length > 0) {
      var thisPos = stack.indexOf(this)
      ~thisPos ? stack.splice(thisPos + 1) : stack.push(this)
      ~thisPos ? keys.splice(thisPos, Infinity, key) : keys.push(key)
      if (~stack.indexOf(value)) value = cycleReplacer.call(this, key, value)
    }
    else stack.push(value)

    return replacer == null ? value : replacer.call(this, key, value)
  }
}
*/
class Ac4yJSONHandler {
    
    isValid(aJSONString){

        try {
            
            var vObject = JSON.parse(aJSONString);
            
            return true;
            
        } catch(exception){return false;}        
        
    } // isValid
    
    serialized(aObject){return JSON.stringify(JSON.decycle(aObject));}
    
    deserialized(aJSONString){return JSON.parse(aJSONString);};
    
    tryDeserialize(aJSONString){
        
        if (this.isValid(aJSONString))
            return JSON.parse(aJSONString);
        else
            return {warning: "érvénytelen objektum"};
    
    }; // tryDeserialize
    
} // Ac4yJSONHandler
/*
            function stringify(obj, replacer, spaces, cycleReplacer) {
              return JSON.stringify(obj, serializer(replacer, cycleReplacer), spaces)
            }

            function serializer(replacer, cycleReplacer) {
              var stack = [], keys = []

              if (cycleReplacer == null) cycleReplacer = function(key, value) {
                if (stack[0] === value) return "[Circular ~]"
                return "[Circular ~." + keys.slice(0, stack.indexOf(value)).join(".") + "]"
              }

              return function(key, value) {
                if (stack.length > 0) {
                  var thisPos = stack.indexOf(this)
                  ~thisPos ? stack.splice(thisPos + 1) : stack.push(this)
                  ~thisPos ? keys.splice(thisPos, Infinity, key) : keys.push(key)
                  if (~stack.indexOf(value)) value = cycleReplacer.call(this, key, value)
                }
                else stack.push(value)

                return replacer == null ? value : replacer.call(this, key, value)
              }
            } 
*/

module.exports=Ac4yJSONHandler
