// @ac4y/ac4y-flow/algebra/Ac4yFlowAlgebra.js
var Ac4yLifecycle=require("@ac4y/ac4y-lifecycle/domain/Ac4yLifecycle.js");
var Ac4yFLCDStock=require("../domain/Ac4yFLCDStock.js");
var Ac4yTimer=require("@ac4y/ac4y-timer/domain/Ac4yTimer.js");

class Ac4yFlowAlgebra extends Ac4yLifecycle {

    constructor(object) {

        super(object);
        if (object != undefined && object === 0) return;
    } // constructor


    createSelf(object) {
        return new Ac4yFlowAlgebra(object);

    } // createSelf


    setDimension(aDimension) {
        this.dimension = aDimension;
    }
    getDimension() {
        return this.dimension;
    }
    createDimension() {
        this.dimension = new Ac4yFLCDStock();
    }
    hasDimension() {
        return this.dimension != undefined;
    }

    setDuration(aDuration) {
        this.duration = aDuration;
    }
    getDuration() {
        return this.duration;
    }
    hasDuration() {
        return this.duration != undefined;
    }

    setDurationInSecond(aDurationInSecond) {
        this.durationInSecond = aDurationInSecond;
    }
    getDurationInSecond() {
        return this.durationInSecond;
    }
    hasDurationInSecond() {
        return this.durationInSecond != undefined;
    }

    setDurationInMinute(aDurationInMinute) {
        this.durationInMinute = aDurationInMinute;
    }
    getDurationInMinute() {
        return this.durationInMinute;
    }
    hasDurationInMinute() {
        return this.durationInMinute != undefined;
    }

    setTimer(aTimer) {
        this.timer = aTimer;
    }
    getTimer() {
        return this.timer;
    }
    createTimer() {
        this.timer = new Ac4yTimer();
    }
    hasTimer() {
        return this.timer != undefined;
    }

    setOnTimeout(aOnTimeout) {
        this.onTimeout = aOnTimeout;
    }
    getOnTimeout() {
        return this.onTimeout;
    }
    hasOnTimeout() {
        return this.onTimeout != undefined;
    }
    _fireOnTimeout() {
        this.onTimeout();
    };
    _tryFireOnTimeout() {
        if (this.onTimeout) this._fireOnTimeout();
    };


    rebuild(object, target) {

        target = target || this;
        super.rebuild(object, target);
        if (object) {

            if (object.dimension != undefined) target.setDimension(new Ac4yFLCDStock(0).rebuilded(object.dimension));
            if (object.duration != undefined) target.setDuration(object.duration);
            if (object.durationInSecond != undefined) target.setDurationInSecond(object.durationInSecond);
            if (object.durationInMinute != undefined) target.setDurationInMinute(object.durationInMinute);
            if (object.timer != undefined) target.setTimer(new Ac4yTimer(0).rebuilded(object.timer));
            if (object.onTimeout != undefined) target.setOnTimeout(object.onTimeout);

        } // if object does not empty

        return target;

    } // rebuild

    rebuilded(object, target) {
        return super.rebuilded(object, target);
    }

    copy(object, target) {

        target = target || this;
        super.copy(object, target);
        if (object) {

            if (object.dimension != undefined) target.setDimension(object.dimension);
            if (object.duration != undefined) target.setDuration(object.duration);
            if (object.durationInSecond != undefined) target.setDurationInSecond(object.durationInSecond);
            if (object.durationInMinute != undefined) target.setDurationInMinute(object.durationInMinute);
            if (object.timer != undefined) target.setTimer(object.timer);
            if (object.onTimeout != undefined) target.setOnTimeout(object.onTimeout);

        } // if object does not empty

        return target;

    } // copy

} // Ac4yFlowAlgebra
		
module.exports=Ac4yFlowAlgebra
