// @ac4y/ac4y-flow/domain/Ac4yFlow.js
var Ac4yFlowAlgebra=require("../algebra/Ac4yFlowAlgebra.js");
var Ac4yLifecycle=require("@ac4y/ac4y-lifecycle/domain/Ac4yLifecycle.js");
var Ac4yTimer=require("@ac4y/ac4y-timer/domain/Ac4yTimer.js");
var Ac4yDateTimeHandler=require("@ac4y/ac4y-utility/domain/Ac4yDateTimeHandler.js");  

class Ac4yFlow extends Ac4yFlowAlgebra {
    
    constructor(object){

        //if (!this.hasStateMachine()) this.createStateMachine();

        super(object);

        if (object != undefined && object === 0) return;

        if (this.hasDurationInSecond()) this.setDuration(new Ac4yDateTimeHandler().getSecondInMillisecond(this.getDurationInSecond()));
        
        if (this.hasDurationInMinute()) this.setDuration(new Ac4yDateTimeHandler().getMinuteInMillisecond(this.getDurationInMinute()));
 
        if (!this.hasDuration()) 
            this.setDuration(3000);

        this.setTimer(
            new Ac4yTimer({ 
                    name: "timeout timer"
                    ,duration: this.getDuration()
                    ,onTime: ( () => {this.timeout();})
            })
        );

        if (!this.getAc4yIdentification().hasTemplate()) this.getAc4yIdentification().createTemplate();
        this.getAc4yIdentification().getTemplate().setHumanId("Ac4yFlow");

    } // constructor    

    createSelf(object) {return new Ac4yFlow(object);}
    
    noLifecycleElements(object){

        object = object || this;

        var object = super.noLifecycleElements(object);

        delete object.timer;
        delete object.duration;
        delete object.durationInSecond;
        delete object.durationInMinute;

        return object;

    } // noLifecycleElements
    
    forTransport(object) {

        object = object || this;

        return object
                //.noIdentification()
                .noLifecycleElements()
                .noStandaloneIdenficationElements()
                .noEcosystem()

    } // forTransport
    
    start(){
    
        //console.log("Ac4yFlow.start");
    
        this.getTimer().start();

        super.start();
        
    } // start

    stop(){
    
        //console.log("Ac4yFlow.stop");
        
        this.getTimer().stop();
        
        super.stop();
        
    } // stop

    timeout(){
    
        //console.log("Ac4yFlow.timeout");
        
        this.getDimension().state().becomesTimeout();
        this.getDimension().stage().reachesIdle();
        this.getDimension().step().realizesTimeout();
        
        this.stop();
        
    } // timeout
    
} // Ac4yFlow

module.exports=Ac4yFlow
