// @ac4y/ac4y-ecosystem/algebra/Ac4yEcosystemMemberAlgebra.js
var Ac4yPersistent=require("@ac4y/ac4y-object/domain/Ac4yPersistent.js");
var Ac4yEcosystem=require("../domain/Ac4yEcosystem.js");

class Ac4yEcosystemMemberAlgebra extends Ac4yPersistent {

    constructor(object) {

        super(object);
        if (object != undefined && object === 0) return;
    } // constructor


    createSelf(object) {
        return new Ac4yEcosystemMemberAlgebra(object);

    } // createSelf


    setEcosystem(aEcosystem) {
        this.ecosystem = aEcosystem;
    }
    getEcosystem() {
        return this.ecosystem;
    }
    createEcosystem() {
        this.ecosystem = new Ac4yEcosystem();
    }
    hasEcosystem() {
        return this.ecosystem != undefined;
    }

    setParent(aParent) {
        this.parent = aParent;
    }
    getParent() {
        return this.parent;
    }
    createParent() {
        this.parent = new Object();
    }
    hasParent() {
        return this.parent != undefined;
    }


    rebuild(object, target) {

        target = target || this;
        super.rebuild(object, target);
        if (object) {

            if (object.ecosystem != undefined) target.setEcosystem(new Ac4yEcosystem(0).rebuilded(object.ecosystem));
            if (object.parent != undefined) target.setParent(new Object(0).rebuilded(object.parent));

        } // if object does not empty

        return target;

    } // rebuild

    rebuilded(object, target) {
        return super.rebuilded(object, target);
    }

    copy(object, target) {

        target = target || this;
        super.copy(object, target);
        if (object) {

            if (object.ecosystem != undefined) target.setEcosystem(object.ecosystem);
            if (object.parent != undefined) target.setParent(object.parent);

        } // if object does not empty

        return target;

    } // copy

} // Ac4yEcosystemMemberAlgebra

module.exports=Ac4yEcosystemMemberAlgebra
