// @ac4y/ac4y-ecosystem/algebra/Ac4yEcosystemAlgebra.js
var Ac4yKeyValueMemory=require("@ac4y/ac4y-common/container/Ac4yKeyValueMemory.js");

class Ac4yEcosystemAlgebra {

    constructor(object) {

        if (object != undefined && object === 0) return;
        if (object) this.copy(object, this);

    } // constructor


    createSelf(object) {
        return new Ac4yEcosystemAlgebra(object);

    } // createSelf


    setMember(aMember) {
        this.member = aMember;
    }
    getMember() {
        return this.member;
    }
    createMember() {
        this.member = new Ac4yKeyValueMemory();
    }
    hasMember() {
        return this.member != undefined;
    }

    setName(aName) {
        this.name = aName;
    }
    getName() {
        return this.name;
    }
    hasName() {
        return this.name != undefined;
    }


    rebuild(object, target) {

        target = target || this;
        if (object) {

            if (object.member != undefined) target.setMember(new Ac4yKeyValueMemory(0).rebuilded(object.member));
            if (object.name != undefined) target.setName(object.name);

        } // if object does not empty

        return target;

    } // rebuild

    rebuilded(object, target) {
        target = target || this;
        if (object) return this.copy(object, this.rebuild(object));

        else return target;

    } // rebuilded

    copy(object, target) {

        target = target || this;
        if (object) {

            if (object.member != undefined) target.setMember(object.member);
            if (object.name != undefined) target.setName(object.name);

        } // if object does not empty

        return target;

    } // copy

} // Ac4yEcosystemAlgebra

module.exports=Ac4yEcosystemAlgebra
