// @ac4y/ac4y-service/domain/Ac4yProcessResult.js
var Ac4yProcessResultAlgebra=require("../algebra/Ac4yProcessResultAlgebra.js");
var Ac4yUserText=require("@ac4y/ac4y-usertext/domain/Ac4yUserText.js"); 

class Ac4yProcessResult extends Ac4yProcessResultAlgebra {

    error(description){

        this.setCode(-1);
        this.setMessage(new Ac4yUserText().text("error!"));
        this.setDescription(description);

        return this;

    } // error

    success(narrative){

        this.setCode(1);

        if (narrative)
            this.setMessage(narrative);
        else
            this.setMessage(new Ac4yUserText().text("success!"));

        return this;

    } // success

    nothingHappened(message, narrative){

        this.setCode(0);

        if (message)
            this.setMessage(message);
        else
            this.setMessage(new Ac4yUserText().text("nothing happened!"));

        if (narrative)
            this.setDescription(narrative);

        return this;

    } // nothingHappened

    itWasFailed(){return this.getCode()==-1;};
    itWasNothingToDo(){return this.getCode()==0;};
    itWasSuccessful(){return this.getCode()==1;};

    addedRequestId(requestId){

        this.setRequestId(requestId);
        return this;

    } // addedRequestId
	
    addedDescription(description){

        this.setDescription(description);

        return this;

    } // addedDescription
        
} // Ac4yProcessResult

module.exports=Ac4yProcessResult
