// @ac4y/ac4y-service/domain/Ac4yServiceResponse.js
var Ac4yServiceResponseAlgebra=require("../algebra/Ac4yServiceResponseAlgebra.js");
var Ac4yProcessResult=require("./Ac4yProcessResult.js");
var Ac4yUserText=require("@ac4y/ac4y-usertext/domain/Ac4yUserText.js"); 

class Ac4yServiceResponse extends Ac4yServiceResponseAlgebra {

    constructor(object) {

        super(object);
        if (object != undefined && object === 0) return;
/*
        if (!this.getAc4yIdentification().hasTemplate()) this.getAc4yIdentification().createTemplate();
        this.getAc4yIdentification().getTemplate().setHumanId("Ac4yServiceResponse");
*/
        if (!this.hasResult()) this.createResult();
        
    } // constructor

    createSelf(object) {return new Ac4yServiceResponse(object);}
    
    getErrorProcessResult(description){return new Ac4yProcessResult().error(description);}

    error(description){

        this.setResult(new Ac4yProcessResult().error(description));
        return this;

    } // error

    success(){

        this.setResult(new Ac4yProcessResult().success());
        return this;

    } // success

    nothingHappened(narrative){

        this.setResult(new Ac4yProcessResult().nothingHappened(
                new Ac4yUserText().text("nothing happened") 
                ,narrative
        ));

        return this;

    } // nothingHappened

    timeout(){return new Ac4yProcessResult().error(new Ac4yUserText().text("timeout!"));}

    getErrorServiceResponse(description){
        return new Ac4yServiceResponse({
            result: this.getErrorProcessResult(description)
        });
    } // getErrorServiceResponse

    getTimeoutServiceResponse(){return new Ac4yProcessResult().error(new Ac4yUserText().text("timeout!"));}
    getSuccessProcessResult(){return new Ac4yProcessResult().success(new Ac4yUserText().text("success"));}
    
    getNothingHappenedProcessResult(narrative){return new Ac4yProcessResult().nothingHappened(
        new Ac4yUserText().text("nothing happened")
        , narrative);
    }

    getSuccessServiceResponse(){
        
        return new Ac4yServiceResponse({
                        result: this.getSuccessProcessResult()
                    });
        
    } // getSuccessServiceResponse


    getNothingHappenedServiceResponse(aNarrativa){
        
        return new Ac4yServiceResponse({
                        result: this.getNothingHappenedProcessResult(aNarrativa)
                    });
        
    } // getNothingHappenedServiceResponse
    	
    itWasFailed(){return this.getResult().getCode()==-1}
    itWasNothingToDo(){return this.getResult().getCode()==0}
    itWasSuccessful(){return this.getResult().getCode()==1}
	
    addedRequestId(requestId){

        this.setResult(this.getResult().addedRequestId(requestId));
        
        return this;

    } // addedRequestId

    addedDescription(description){

        this.setResult(this.getResult().addedDescription(description));

        return this;

    } // addedDescription
        
} // Ac4yServiceResponse
	
module.exports=Ac4yServiceResponse
