// @ac4y/ac4y-service/domain/Ac4yObjectService.js
var Ac4yServiceResponse=require("./Ac4yServiceResponse.js");
var Ac4yProcessResult=require("./Ac4yProcessResult.js");

var GetObjectResponse = require("../document/GetObjectResponse.js");
var GetListResponse = require("../document/GetListResponse.js");

class Ac4yObjectService extends Ac4yServiceResponse {

    async getList(list, name) {

        try {

            var response = new GetListResponse();

            if (name)
                response[name]=list;
            else
                response.setList(list);

            if (response.hasList() && response.getList().length>0)
                response.setResult(new Ac4yProcessResult().success());
            else
                response.setResult(new Ac4yProcessResult().nothingHappened("the list is empty"));

        } catch (error) {response.setResult(new Ac4yProcessResult().error(error.message || error))};

        return response;

    } // getList

    async getObject(object) {

        try {

            var response = new GetObjectResponse();

            if (object) {

                response.setObject(object);
                response.setResult(new Ac4yProcessResult().success());

            }
            else
                response.setResult(new Ac4yProcessResult().nothingHappened("has no result!"));

        } catch (error) {response.setResult(new Ac4yProcessResult().error(error.message || error))};

        return response;

    } // getObject

    async getObjectResponse(service) {

        try {

            var response = new GetObjectResponse();

            var object = await service.catch( (error) => {throw error});

            if (object) {

                response.setObject(object);
                response.setResult(new Ac4yProcessResult().success());

            }
            else
                response.setResult(new Ac4yProcessResult().nothingHappened("has no result!"));

        } catch (error) {response.setResult(new Ac4yProcessResult().error(error.message || error))};

        return response;

    } // getObjectResponse

    async doesExistObjectById(service) {

        try {

            var response = new GetObjectResponse();

            var exists = await service.catch( (error) => {throw error});

            if (exists)
                response.setResult(
                    new Ac4yProcessResult({
                        code: 1
                        ,message: "exist!"
                    })
                );
            else
                response.setResult(
                    new Ac4yProcessResult({
                        code: 0
                        ,message: "does not exist!"
                    })
                );

        } catch (error) {response.setResult(new Ac4yProcessResult().error(error.message || error))};

        return response;

    } // doesExistObjectById
	
} // Ac4yObjectService 

module.exports=Ac4yObjectService
