// @ac4y/ac4y-service/domain/Ac4yService.js
var Ac4yServiceAlgebra=require("../algebra/Ac4yServiceAlgebra.js");
var Ac4yServiceResponse=require("./Ac4yServiceResponse.js");
var Ac4yProcessResult=require("./Ac4yProcessResult.js");
var Ac4yObjectHandler=require("@ac4y/ac4y-utility/domain/Ac4yObjectHandler.js");

class Ac4yService extends Ac4yServiceAlgebra {

    constructor(object) {

        super(object);
        
        if (object != undefined && object === 0) return;     

        if (!this.hasDimension()) this.createDimension();

        if (!this.getAc4yIdentification().hasTemplate()) this.getAc4yIdentification().createTemplate();
        this.getAc4yIdentification().getTemplate().setHumanId("Ac4yService");

    } // constructor

    createSelf(object) {return new Ac4yService(object);}
    
    lightweight(){

        var lightweight = this.createSelf(this);

        //delete lightweight.getAc4yIdentification().template;
        delete lightweight.getAc4yIdentification().createdAt;
        delete lightweight.timer;
        delete lightweight.liveIndicator;
        delete lightweight.duration;
        delete lightweight.durationInSecond;
        delete lightweight.dimension;
        delete lightweight.startTime;
        delete lightweight.GUID;
        delete lightweight.createdAt;

        return lightweight;

    } // lightweight
    
    forPersistence(){

        var object = this.createSelf(this);
        
        if (object.hasRequest() && object.getRequest().hasEnvelope() && object.getRequest().getEnvelope().hasOwner())
            object.setOwner(object.getRequest().getEnvelope().getOwner())

        object.check();

        object = this.objectKilledForPersistence(object);
        object = super.forPersistence(object);        
        
        delete object.duration;
        delete object.durationInSecond;

        object.startTime=new Date(object.startTime);
        object.stopTime=new Date(object.stopTime);

        return object;

    } // forPersistence     
    fromPersistence(){

        this.getAc4yIdentification().setGUID(this.getGUID());

        return this;

    } // fromPersistence
    
    getErrorProcessResult(description){return new Ac4yServiceResponse().getErrorProcessResult(description)};
    error(description){return new Ac4yServiceResponse().error(description)};

    getErrorServiceResponse(aErrorDescription){
        
        return new Ac4yServiceResponse({
                        result: this.getErrorProcessResult(aErrorDescription)
                    });
        
    } // getErrorServiceResponse

    getTimeoutServiceResponse(){return this.getErrorServiceResponse("timeout");}
    
    getSuccessProcessResult(){return new Ac4yServiceResponse().getSuccessProcessResult()};

    getSuccessServiceResponse(){
        
        return new Ac4yServiceResponse({
                        result: this.getSuccessProcessResult()
                    });
        
    } // getSuccessServiceResponse

    success(){return new Ac4yServiceResponse().success()};

    getNothingHappenedProcessResult(narrative){return new Ac4yServiceResponse().getNothingHappenedProcessResult(narrative)}

    getNothingHappenedServiceResponse(narrative){
        
        return new Ac4yServiceResponse({
                        result: this.getNothingHappenedProcessResult(narrative)
                    });
        
    } // getNothingHappenedServiceResponse

    nothingHappened(narrative){return new Ac4yServiceResponse().nothingHappened(narrative)}

    done(){

//        console.log("Ac4yService.done");
        
        this.getDimension().state().becomesDone();
        
        this._tryFireOnDone();
        
        switch (new Ac4yServiceResponse(0).rebuilded(this.getResponse()).getResult().getCode()) {
            
            case 1: this.success(); break;
            case 0: this.notingHappened(); break;
            case -1: this.fail(); break;
            default: this.fail(); 
                
        } // switch
        
    } // done

    rollback(){

//        console.log("Ac4yService.rollback");
        
        this.getDimension().state().becomesRollback();
        
        this._tryFireOnRollback();
        
    } // rollback

    commit(){

//        console.log("Ac4yService.commit");
        
        this.getDimension().state().becomesCommit();
        
        this._tryFireOnCommit();
        
    } // commit
    
    success(){

//        console.log("Ac4yService.success");
        
        this.getDimension().state().becomesSuccess();

        this._tryFireOnSuccess();

        this.commit();
        
        this.stop();
        
    } // success
    
    fail(){

//        console.log("Ac4yService.fail");
        
        this.getDimension().state().becomesFail();

        this._tryFireOnFail();

        this.rollback();
        
        this.stop();

    } // fail
    
    nothingHappened(){

//        console.log("Ac4yService.nothingHappened");
        
        this.getDimension().state().becomesNothingHappened();

        this._tryFireOnNothingHappened();

        this.rollback();
        
        this.stop();

    } // nothingHappened
    
    timeout(){
    
//        console.log("Ac4yService.timeout");
        
        this.getDimension().state().becomesTimeout();

        this._tryFireOnTimeout();

        //this.rollback();
        
        this.stop();
        
    } // timeout
        
    getResult(){return new Ac4yServiceResponse();}

}; // Ac4yService

module.exports=Ac4yService
