// @ac4y/ac4y-lifecycle/algebra/Ac4yLifecycleAlgebra.js
var Ac4yEcosystemMember=require("@ac4y/ac4y-ecosystem/domain/Ac4yEcosystemMember.js"); 
var Ac4yBulb=require("@ac4y/ac4y-common/domain/Ac4yBulb.js");
var Ac4yArray=require("@ac4y/ac4y-common/container/Ac4yArray.js");
var Ac4yLCDStock=require("../domain/Ac4yLCDStock.js");
var Ac4yKeyValueMemory=require("@ac4y/ac4y-common/container/Ac4yKeyValueMemory.js");

class Ac4yLifecycleAlgebra extends Ac4yEcosystemMember {

    constructor(object) {

        super(object);
        if (object != undefined && object === 0) return;
    } // constructor


    createSelf(object) {
        return new Ac4yLifecycleAlgebra(object);

    } // createSelf


    setDimension(aDimension) {
        this.dimension = aDimension;
    }
    getDimension() {
        return this.dimension;
    }
    createDimension() {
        this.dimension = new Ac4yLCDStock();
    }
    hasDimension() {
        return this.dimension != undefined;
    }

    setIndicatorSet(aIndicatorSet) {
        this.indicatorSet = aIndicatorSet;
    }
    getIndicatorSet() {
        return this.indicatorSet;
    }
    createIndicatorSet() {
        this.indicatorSet = new Ac4yKeyValueMemory();
    }
    hasIndicatorSet() {
        return this.indicatorSet != undefined;
    }

    setSubset(aSubset) {
        this.subset = aSubset;
    }
    getSubset() {
        return this.subset;
    }
    createSubset() {
        this.subset = new Ac4yArray();
    }
    hasSubset() {
        return this.subset != undefined;
    }

    setState(aState) {
        this.state = aState;
    }
    getState() {
        return this.state;
    }
    hasState() {
        return this.state != undefined;
    }

    setStep(aStep) {
        this.step = aStep;
    }
    getStep() {
        return this.step;
    }
    hasStep() {
        return this.step != undefined;
    }

    setStage(aStage) {
        this.stage = aStage;
    }
    getStage() {
        return this.stage;
    }
    hasStage() {
        return this.stage != undefined;
    }

    setStation(aStation) {
        this.station = aStation;
    }
    getStation() {
        return this.station;
    }
    hasStation() {
        return this.station != undefined;
    }

    setStartTime(aStartTime) {
        this.startTime = aStartTime;
    }
    getStartTime() {
        return this.startTime;
    }
    hasStartTime() {
        return this.startTime != undefined;
    }

    setStopTime(aStopTime) {
        this.stopTime = aStopTime;
    }
    getStopTime() {
        return this.stopTime;
    }
    hasStopTime() {
        return this.stopTime != undefined;
    }

    setLiveIndicator(aLiveIndicator) {
        this.liveIndicator = aLiveIndicator;
    }
    getLiveIndicator() {
        return this.liveIndicator;
    }
    createLiveIndicator() {
        this.liveIndicator = new Ac4yBulb();
    }
    hasLiveIndicator() {
        return this.liveIndicator != undefined;
    }

    setOnCreate(aOnCreate) {
        this.onCreate = aOnCreate;
    }
    getOnCreate() {
        return this.onCreate;
    }
    hasOnCreate() {
        return this.onCreate != undefined;
    }
    _fireOnCreate() {
        this.onCreate();
    };
    _tryFireOnCreate() {
        if (this.onCreate) this._fireOnCreate();
    };

    setOnBorn(aOnBorn) {
        this.onBorn = aOnBorn;
    }
    getOnBorn() {
        return this.onBorn;
    }
    hasOnBorn() {
        return this.onBorn != undefined;
    }
    _fireOnBorn() {
        this.onBorn();
    };
    _tryFireOnBorn() {
        if (this.onBorn) this._fireOnBorn();
    };

    setOnDeath(aOnDeath) {
        this.onDeath = aOnDeath;
    }
    getOnDeath() {
        return this.onDeath;
    }
    hasOnDeath() {
        return this.onDeath != undefined;
    }
    _fireOnDeath() {
        this.onDeath();
    };
    _tryFireOnDeath() {
        if (this.onDeath) this._fireOnDeath();
    };

    setOnStart(aOnStart) {
        this.onStart = aOnStart;
    }
    getOnStart() {
        return this.onStart;
    }
    hasOnStart() {
        return this.onStart != undefined;
    }
    _fireOnStart() {
        this.onStart();
    };
    _tryFireOnStart() {
        if (this.onStart) this._fireOnStart();
    };

    setOnStop(aOnStop) {
        this.onStop = aOnStop;
    }
    getOnStop() {
        return this.onStop;
    }
    hasOnStop() {
        return this.onStop != undefined;
    }
    _fireOnStop() {
        this.onStop();
    };
    _tryFireOnStop() {
        if (this.onStop) this._fireOnStop();
    };

    setOnPause(aOnPause) {
        this.onPause = aOnPause;
    }
    getOnPause() {
        return this.onPause;
    }
    hasOnPause() {
        return this.onPause != undefined;
    }
    _fireOnPause() {
        this.onPause();
    };
    _tryFireOnPause() {
        if (this.onPause) this._fireOnPause();
    };

    setOnContinue(aOnContinue) {
        this.onContinue = aOnContinue;
    }
    getOnContinue() {
        return this.onContinue;
    }
    hasOnContinue() {
        return this.onContinue != undefined;
    }
    _fireOnContinue() {
        this.onContinue();
    };
    _tryFireOnContinue() {
        if (this.onContinue) this._fireOnContinue();
    };

    setOnError(aOnError) {
        this.onError = aOnError;
    }
    getOnError() {
        return this.onError;
    }
    hasOnError() {
        return this.onError != undefined;
    }
    _fireOnError() {
        this.onError();
    };
    _tryFireOnError() {
        if (this.onError) this._fireOnError();
    };


    rebuild(object, target) {

        target = target || this;
        super.rebuild(object, target);
        if (object) {

            if (object.dimension != undefined) target.setDimension(new Ac4yLCDStock(0).rebuilded(object.dimension));
            if (object.indicatorSet != undefined) target.setIndicatorSet(new Ac4yKeyValueMemory(0).rebuilded(object.indicatorSet));
            if (object.subset != undefined) target.setSubset(new Ac4yArray(0).rebuilded(object.subset));
            if (object.state != undefined) target.setState(object.state);
            if (object.step != undefined) target.setStep(object.step);
            if (object.stage != undefined) target.setStage(object.stage);
            if (object.station != undefined) target.setStation(object.station);
            if (object.startTime != undefined) target.setStartTime(object.startTime);
            if (object.stopTime != undefined) target.setStopTime(object.stopTime);
            if (object.liveIndicator != undefined) target.setLiveIndicator(new Ac4yBulb(0).rebuilded(object.liveIndicator));
            if (object.onCreate != undefined) target.setOnCreate(object.onCreate);
            if (object.onBorn != undefined) target.setOnBorn(object.onBorn);
            if (object.onDeath != undefined) target.setOnDeath(object.onDeath);
            if (object.onStart != undefined) target.setOnStart(object.onStart);
            if (object.onStop != undefined) target.setOnStop(object.onStop);
            if (object.onPause != undefined) target.setOnPause(object.onPause);
            if (object.onContinue != undefined) target.setOnContinue(object.onContinue);
            if (object.onError != undefined) target.setOnError(object.onError);

        } // if object does not empty

        return target;

    } // rebuild

    rebuilded(object, target) {
        return super.rebuilded(object, target);
    }

    copy(object, target) {

        target = target || this;
        super.copy(object, target);
        if (object) {

            if (object.dimension != undefined) target.setDimension(object.dimension);
            if (object.indicatorSet != undefined) target.setIndicatorSet(object.indicatorSet);
            if (object.subset != undefined) target.setSubset(object.subset);
            if (object.state != undefined) target.setState(object.state);
            if (object.step != undefined) target.setStep(object.step);
            if (object.stage != undefined) target.setStage(object.stage);
            if (object.station != undefined) target.setStation(object.station);
            if (object.startTime != undefined) target.setStartTime(object.startTime);
            if (object.stopTime != undefined) target.setStopTime(object.stopTime);
            if (object.liveIndicator != undefined) target.setLiveIndicator(object.liveIndicator);
            if (object.onCreate != undefined) target.setOnCreate(object.onCreate);
            if (object.onBorn != undefined) target.setOnBorn(object.onBorn);
            if (object.onDeath != undefined) target.setOnDeath(object.onDeath);
            if (object.onStart != undefined) target.setOnStart(object.onStart);
            if (object.onStop != undefined) target.setOnStop(object.onStop);
            if (object.onPause != undefined) target.setOnPause(object.onPause);
            if (object.onContinue != undefined) target.setOnContinue(object.onContinue);
            if (object.onError != undefined) target.setOnError(object.onError);

        } // if object does not empty

        return target;

    } // copy

} // Ac4yLifecycleAlgebra
		
module.exports=Ac4yLifecycleAlgebra
