// @ac4y/ac4y-lifecycle/algebra/Ac4yLifecycleDimensionAlgebra.js
var Ac4yPersistent=require("@ac4y/ac4y-object/domain/Ac4yPersistent.js");
var Ac4yKeyValueMemory=require("@ac4y/ac4y-common/container/Ac4yKeyValueMemory.js"); 
var Ac4yArray=require("@ac4y/ac4y-common/container/Ac4yArray.js");

class Ac4yLifecycleDimensionAlgebra extends Ac4yPersistent {

    constructor(object) {

        super(object);
        if (object != undefined && object === 0) return;
    } // constructor


    createSelf(object) {
        return new Ac4yLifecycleDimensionAlgebra(object);

    } // createSelf


    setDimension(aDimension) {
        this.dimension = aDimension;
    }
    getDimension() {
        return this.dimension;
    }
    hasDimension() {
        return this.dimension != undefined;
    }

    setFact(aFact) {
        this.fact = aFact;
    }
    getFact() {
        return this.fact;
    }
    hasFact() {
        return this.fact != undefined;
    }

    setStock(aStock) {
        this.stock = aStock;
    }
    getStock() {
        return this.stock;
    }
    createStock() {
        this.stock = new Ac4yKeyValueMemory();
    }
    hasStock() {
        return this.stock != undefined;
    }

    setHistory(aHistory) {
        this.history = aHistory;
    }
    getHistory() {
        return this.history;
    }
    createHistory() {
        this.history = new Ac4yArray();
    }
    hasHistory() {
        return this.history != undefined;
    }


    rebuild(object, target) {

        target = target || this;
        super.rebuild(object, target);
        if (object) {

            if (object.dimension != undefined) target.setDimension(object.dimension);
            if (object.fact != undefined) target.setFact(object.fact);
            if (object.stock != undefined) target.setStock(new Ac4yKeyValueMemory(0).rebuilded(object.stock));
            if (object.history != undefined) target.setHistory(new Ac4yArray(0).rebuilded(object.history));

        } // if object does not empty

        return target;

    } // rebuild

    rebuilded(object, target) {
        return super.rebuilded(object, target);
    }

    copy(object, target) {

        target = target || this;
        super.copy(object, target);
        if (object) {

            if (object.dimension != undefined) target.setDimension(object.dimension);
            if (object.fact != undefined) target.setFact(object.fact);
            if (object.stock != undefined) target.setStock(object.stock);
            if (object.history != undefined) target.setHistory(object.history);

        } // if object does not empty

        return target;

    } // copy

} // Ac4yLifecycleDimensionAlgebra
		
module.exports=Ac4yLifecycleDimensionAlgebra
