// @ac4y/ac4y-lifecycle/algebra/Ac4yLifecycleFactAlgebra.js
var Ac4yPersistent=require("@ac4y/ac4y-object/domain/Ac4yPersistent.js");

class Ac4yLifecycleFactAlgebra extends Ac4yPersistent {

    constructor(object) {

        super(object);
        if (object != undefined && object === 0) return;
    } // constructor


    createSelf(object) {
        return new Ac4yLifecycleFactAlgebra(object);

    } // createSelf


    setDimension(aDimension) {
        this.dimension = aDimension;
    }
    getDimension() {
        return this.dimension;
    }
    hasDimension() {
        return this.dimension != undefined;
    }

    setName(aName) {
        this.name = aName;
    }
    getName() {
        return this.name;
    }
    hasName() {
        return this.name != undefined;
    }

    setEntry(aEntry) {
        this.entry = aEntry;
    }
    getEntry() {
        return this.entry;
    }
    hasEntry() {
        return this.entry != undefined;
    }

    setQuit(aQuit) {
        this.quit = aQuit;
    }
    getQuit() {
        return this.quit;
    }
    hasQuit() {
        return this.quit != undefined;
    }

    setStay(aStay) {
        this.stay = aStay;
    }
    getStay() {
        return this.stay;
    }
    hasStay() {
        return this.stay != undefined;
    }

    setOrigin(aOrigin) {
        this.origin = aOrigin;
    }
    getOrigin() {
        return this.origin;
    }
    hasOrigin() {
        return this.origin != undefined;
    }


    rebuild(object, target) {

        target = target || this;
        super.rebuild(object, target);
        if (object) {

            if (object.dimension != undefined) target.setDimension(object.dimension);
            if (object.name != undefined) target.setName(object.name);
            if (object.entry != undefined) target.setEntry(object.entry);
            if (object.quit != undefined) target.setQuit(object.quit);
            if (object.stay != undefined) target.setStay(object.stay);
            if (object.origin != undefined) target.setOrigin(object.origin);

        } // if object does not empty

        return target;

    } // rebuild

    rebuilded(object, target) {
        return super.rebuilded(object, target);
    }

    copy(object, target) {

        target = target || this;
        super.copy(object, target);
        if (object) {

            if (object.dimension != undefined) target.setDimension(object.dimension);
            if (object.name != undefined) target.setName(object.name);
            if (object.entry != undefined) target.setEntry(object.entry);
            if (object.quit != undefined) target.setQuit(object.quit);
            if (object.stay != undefined) target.setStay(object.stay);
            if (object.origin != undefined) target.setOrigin(object.origin);

        } // if object does not empty

        return target;

    } // copy

} // Ac4yLifecycleFactAlgebra
		
module.exports=Ac4yLifecycleFactAlgebra
