// @ac4y/ac4y-timer/algebra/Ac4yTimerAlgebra.js
var Ac4y=require("@ac4y/ac4y-object/domain/Ac4y.js"); 
var Ac4yBulb=require("@ac4y/ac4y-common/domain/Ac4yBulb.js"); 
var Ac4yDateTimeHandler=require("@ac4y/ac4y-utility/domain/Ac4yDateTimeHandler.js"); 
var Ac4yLifecycle=require("@ac4y/ac4y-lifecycle/domain/Ac4yLifecycle.js");

class Ac4yTimerAlgebra extends Ac4yLifecycle {

    constructor(object) {

        super(object);
        if (object != undefined && object === 0) return;
    } // constructor


    createSelf(object) {
        return new Ac4yTimerAlgebra(object);

    } // createSelf


    setDuration(aDuration) {
        this.duration = aDuration;
    }
    getDuration() {
        return this.duration;
    }
    hasDuration() {
        return this.duration != undefined;
    }

    setDurationInSecond(aDurationInSecond) {
        this.durationInSecond = aDurationInSecond;
    }
    getDurationInSecond() {
        return this.durationInSecond;
    }
    hasDurationInSecond() {
        return this.durationInSecond != undefined;
    }

    setDurationInMinute(aDurationInMinute) {
        this.durationInMinute = aDurationInMinute;
    }
    getDurationInMinute() {
        return this.durationInMinute;
    }
    hasDurationInMinute() {
        return this.durationInMinute != undefined;
    }

    setEndTime(aEndTime) {
        this.endTime = aEndTime;
    }
    getEndTime() {
        return this.endTime;
    }
    hasEndTime() {
        return this.endTime != undefined;
    }

    setTimer(aTimer) {
        this.timer = aTimer;
    }
    getTimer() {
        return this.timer;
    }
    hasTimer() {
        return this.timer != undefined;
    }

    setOnTime(aOnTime) {
        this.onTime = aOnTime;
    }
    getOnTime() {
        return this.onTime;
    }
    hasOnTime() {
        return this.onTime != undefined;
    }
    _fireOnTime() {
        this.onTime();
    };
    _tryFireOnTime() {
        if (this.onTime) this._fireOnTime();
    };

    setOnCancel(aOnCancel) {
        this.onCancel = aOnCancel;
    }
    getOnCancel() {
        return this.onCancel;
    }
    hasOnCancel() {
        return this.onCancel != undefined;
    }
    _fireOnCancel() {
        this.onCancel();
    };
    _tryFireOnCancel() {
        if (this.onCancel) this._fireOnCancel();
    };


    rebuild(object, target) {

        target = target || this;
        super.rebuild(object, target);
        if (object) {

            if (object.duration != undefined) target.setDuration(object.duration);
            if (object.durationInSecond != undefined) target.setDurationInSecond(object.durationInSecond);
            if (object.durationInMinute != undefined) target.setDurationInMinute(object.durationInMinute);
            if (object.endTime != undefined) target.setEndTime(object.endTime);
            if (object.timer != undefined) target.setTimer(object.timer);
            if (object.onTime != undefined) target.setOnTime(object.onTime);
            if (object.onCancel != undefined) target.setOnCancel(object.onCancel);

        } // if object does not empty

        return target;

    } // rebuild

    rebuilded(object, target) {
        return super.rebuilded(object, target);
    }

    copy(object, target) {

        target = target || this;
        super.copy(object, target);
        if (object) {

            if (object.duration != undefined) target.setDuration(object.duration);
            if (object.durationInSecond != undefined) target.setDurationInSecond(object.durationInSecond);
            if (object.durationInMinute != undefined) target.setDurationInMinute(object.durationInMinute);
            if (object.endTime != undefined) target.setEndTime(object.endTime);
            if (object.timer != undefined) target.setTimer(object.timer);
            if (object.onTime != undefined) target.setOnTime(object.onTime);
            if (object.onCancel != undefined) target.setOnCancel(object.onCancel);

        } // if object does not empty

        return target;

    } // copy

} // Ac4yTimerAlgebra

module.exports=Ac4yTimerAlgebra
