// @ac4y/ac4y-service/constant/Ac4yServiceConstant.js
class Ac4yServiceConstant {
    
    constructor() {

        this.REQUEST = "request";
        this.RESPONSE = "response";
        this.COMMIT = "commit";
        this.ROLLBACK = "rollback";
        this.SUCCESS = "success";
        this.FAIL = "fail";
        this.DONE = "done";
        this.NOTHINGHAPPENED = "nothing happened";

        this.STATE = {
            done: "done"
            ,waiting: "waiting"
            ,processing: "processing"
            ,rollback: "rollback"
            ,transporting: "transporting"
            ,request: "request"
            ,response: "response"
            ,call: "call"
            ,reply: "reply"
            ,commit: "commit"
            ,success: "success"
            ,fail: "fail"
            ,warning: "warning"
            ,nothingHappened: "nothing happened"
        };

        this.STEP = {
            done: "done"
            ,request: "request"
            ,response: "response"
            ,process: "process"
            ,transport: "transport"
            ,call: "call"
            ,reply: "reply"
            ,rollback: "rollback"
            ,commit: "commit"
            ,success: "success"
            ,fail: "fail"
            ,warning: "warning"
            ,nothingHappened: "nothing happened"
        };
        
        this.STAGE = {
            waiting4Response: "waiting for response"
        };
        
    } // constructor
    
} // Ac4yServiceConstant

module.exports=Ac4yServiceConstant
